dbt_common/clients/_jinja_blocks.py

```diff
@@ -129,8 +143,33 @@ def advance(self, new_position: int) -> None:
     def rewind(self, amount: int = 1) -> None:
         self.pos -= amount
 
-    def _search(self, pattern: re.Pattern) -> Optional[re.Match]:
-        return pattern.search(self.text, self.pos)
+    def _search(self, pattern) -> Optional[re.Match]:
+
+        # Check to see if we have a cached search on this pattern.
+        positioned_match = self._past_matches.get(pattern)
+
+        if positioned_match is None or positioned_match.start_pos > self.pos:
+            # We did not have a cached search, or we did, but it was done at a location
+            # further along in the string. Do a new search and cache it.
+            match = pattern.search(self.text, self.pos)
+            self._past_matches[pattern] = PositionedMatch(self.pos, match)
+        else:
+            # We have a cached search and its start position falls before (or at) the
+            # current search position...
+            if positioned_match.match is None:
+                # ...but there is no match in the rest of the 'data'.
+                match = None
```

Comment:
Hmm, this can never be `true`, due to the `if positioned_match is None or ...` in the `if` statement above? 🤔 Propose to remove this if-branch.